share/man/man3/sysctl.3
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)sysctl.3	8.1.1 (2.11BSD GTE) 1/13/95
.\"
.TH SYSCTL 3 "January 13, 1995"
.UC 4
.SH NAME
sysctl \- get or set system information
.SH SYNOPSIS
.nf
.ft B
#include <sys/sysctl.h>

int
sysctl(name, namelen, oldp, *oldlenp, *newp, newlen)
   int *name;
   u_int namelen;
   void *oldp;
   size_t *oldlenp;
   void *newp;
   size_t newlen;
.ft R
.fi
.SH DESCRIPTION
The
\fBsysctl\fP
function retrieves system information and allows processes with
appropriate privileges to set system information.
The information available from
\fBsysctl\fP
consists of integers, strings, and tables.
Information may be retrieved and set from the command interface
using the 
\fIsysctl\fP(1)
utility.
.PP
Unless explicitly noted below,
\fBsysctl\fP
returns a consistent snapshot of the data requested.
Calls to
\fBsysctl\fP
are serialized to avoid deadlock.
.PP
The state is described using a ``Management Information Base'' (MIB)
style name, listed in
\fIname\fP,
which is a
\fInamelen\fP
length array of integers.
.PP
The information is copied into the buffer specified by
\fIoldp\fP.
The size of the buffer is given by the location specified by
\fIoldlenp\fP
before the call,
and that location gives the amount of data copied after a successful call.
If the amount of data available is greater
than the size of the buffer supplied,
the call supplies as much data as fits in the buffer provided
and returns with the error code ENOMEM.
If the old value is not desired,
\fIoldp\fP
and
\fIoldlenp\fP
should be set to NULL.
.PP
The size of the available data can be determined by calling 
\fBsysctl\fP
with a NULL parameter for
\fIoldp\fP.
The size of the available data will be returned in the location pointed to by
\fIoldlenp\fP.
For some operations, the amount of space may change often.
For these operations,
the system attempts to round up so that the returned size is
large enough for a call to return the data shortly thereafter.
.PP
To set a new value,
\fInewp\fP
is set to point to a buffer of length
\fInewlen\fP
from which the requested value is to be taken.
If a new value is not to be set,
\fInewp\fP
should be set to NULL and
\fInewlen\fP
set to 0.
.PP
The top level names are defined with a CTL_ prefix in
\fI<sys/sysctl.h>\fP,
and are as follows.
The next and subsequent levels down are found in the include files
listed here, and described in separate sections below.
.PP
.nf
.ta .25i 1.5i 3.75i 4.5i
	Name	Next level names	Description
	CTL\_DEBUG	sys/sysctl.h	Debugging
	CTL\_FS	sys/sysctl.h	File system
	CTL\_HW	sys/sysctl.h	Generic CPU, I/O
	CTL\_KERN	sys/sysctl.h	High kernel limits
	CTL\_MACHDEP	sys/sysctl.h	Machine dependent
	CTL\_NET	sys/socket.h	Networking
	CTL\_USER	sys/sysctl.h	User-level
	CTL\_VM	vm/vm_param.h	Virtual memory
.fi
.PP
For example, the following retrieves the maximum number of processes allowed
in the system:
.sp
.nf
.in +.5i
int mib[2], maxproc;
size_t len;
.sp
mib[0] = CTL_KERN;
mib[1] = KERN_MAXPROC;
len = sizeof(maxproc);
sysctl(mib, 2, &maxproc, &len, NULL, 0);
.in -.5i
.sp
To retrieve the standard search path for the system utilities:
.sp
.nf
.in +.5i
int mib[2];
size_t len;
char *p;

mib[0] = CTL_USER;
mib[1] = USER_CS_PATH;
sysctl(mib, 2, NULL, &len, NULL, 0);
p = malloc(len);
sysctl(mib, 2, p, &len, NULL, 0);
.in -.5i
.fi
.SH CTL_DEBUG
The debugging variables vary from system to system.
A debugging variable may be added or deleted without need to recompile
\fBsysctl\fP
to know about it.
Each time it runs,
\fBsysctl\fP
gets the list of debugging variables from the kernel and
displays their current values.
The system defines twenty 
\fIstruct ctldebug\fP
variables named 
\fIdebug0\fP
through
\fIdebug19\fP.
They are declared as separate variables so that they can be
individually initialized at the location of their associated variable.
The loader prevents multiple use of the same variable by issuing errors
if a variable is initialized in more than one place.
For example, to export the variable
\fIdospecialcheck\fP
as a debugging variable, the following declaration would be used:
.sp
.nf
.in +.5i
int dospecialcheck = 1;
struct ctldebug debug5 = { "dospecialcheck", &dospecialcheck };
.in -.5i
.fi
.SH CTL_FS
There are currently no second level names for the file system.
.SH CTL_HW
The string and integer information available for the CTL_HW level
is detailed below.
The changeable column shows whether a process with appropriate
privilege may change the value.
.sp
.nf
.ta .25i 2.75i 3.75i
	Second level name	Type	Changeable
	HW\_MACHINE	string	no
	HW\_MODEL	string	no
	HW\_NCPU	integer	no
	HW\_BYTEORDER	integer	no
	HW\_PHYSMEM	integer	no
	HW\_USERMEM	integer	no
	HW\_PAGESIZE	integer	no
.\".It HW\_DISKNAMES	integer	no
.\".It HW\_DISKSTATS	integer	no
	HW\_MACHINE_ARCH	string	no
.fi
.PP
.ta .5i
HW_MACHINE
.br
	The machine class.
.sp
HW_MODEL
.br
	The machine model.
.sp
HW_NCPU
.br
	The number of cpus.
.sp
HW_BYTEORDER
.br
	The byteorder (3412, 4321, or 1234).
.sp
HW_PHYSMEM
.br
	The bytes of physical memory.
.sp
HW_USERMEM
.br
	The bytes of non-kernel memory.
.sp
HW_PAGESIZE
.br
	The software page size.
.sp
.\".It Fa HW_DISKNAMES
.\".It Fa HW_DISKSTATS
HW_MACHINE_ARCH
.br
	The machine architecture.
.sp
.SH CTL_KERN
The string and integer information available for the CTL_KERN level
is detailed below.
The changeable column shows whether a process with appropriate
privilege may change the value.
The types of data currently available are process information,
system inodes, the open file entries, routing table entries,
virtual memory statistics, load average history, and clock rate
information.
.sp
.nf
.ta .25i 2.75i 4.5i

	Second level name	Type	Changeable
	KERN\_ARGMAX	integer	no
	KERN\_BOOTTIME	struct timeval	no
	KERN\_CHOWN\_RESTRICTED	integer	no
	KERN\_CLOCKRATE	struct clockinfo	no
	KERN\_FILE	struct file	no
	KERN\_HOSTID	long	yes
	KERN\_HOSTNAME	string	yes
	KERN\_JOB\_CONTROL	integer	no
.\"KERN\_LINK\_MAX	integer	no
	KERN\_MAXFILES	integer	no
	KERN\_MAXPROC	integer	no
	KERN\_MAXINODES	integer	no
	KERN\_MAXTEXTS	integer	no
.\"KERN\_MAX\_CANON	integer	no
.\"KERN\_MAX\_INPUT	integer	no
.\"KERN\_NAME\_MAX	integer	no
	KERN\_NGROUPS	integer	no
.\"KERN\_NO\_TRUNC	integer	no
	KERN\_OSRELEASE	string	no
	KERN\_OSREV	integer	no
	KERN\_OSTYPE	string	no
	KERN\_OSVERSION	string	no
.\"KERN\_PATH\_MAX	integer	no
.\"KERN\_PIPE\_BUF	integer	no
	KERN\_POSIX1	integer	no
	KERN\_PROC	struct proc	no
	KERN\_PROF	node	not applicable
	KERN\_SAVED\_IDS	integer	no
	KERN\_SECURELVL	integer	raise only
	KERN\_TEXT	struct text	no
.\"KERN\_VDISABLE	integer	no
	KERN\_VERSION	string	no
	KERN\_INODE	struct inode	no
.fi
.PP
KERN_ARGMAX
.br
.in +.5i
The maximum bytes of argument to \fBexec(2)\fP.
.sp
.in -.5i
KERN_BOOTTIME
.br
.in +.5i
A \fIstruct timeval\fP structure is returned.
This structure contains the time that the system was booted.
.sp
.in -.5i
.\"KERN_CHOWN_RESTRICTED
.\"Return 1 if appropriate privileges are required for the
.\".Xr chown 2
.\"system call, otherwise 0.
KERN_CLOCKRATE
.br
.in +.5i
A \fIstruct clockinfo\fP structure is returned.
This structure contains the clock, statistics clock and profiling clock
frequencies, and the number of micro-seconds per hz tick.
.sp
.in -.5i
KERN_FILE
.br
.in +.5i
Return the entire file table as an array of extended \fIfile\fP
structures.
Each element of the array contains the kernel address of a file
\fIstruct inode *\fP
followed by the file itself
\fIstruct file\fP.  There can never be more than \fBKERN_MAXFILES\fP
inodes returned.
.sp
.in -.5i
KERN_HOSTID
.br
.in +.5i
Get or set the host id.
.sp
.in -.5i
KERN_HOSTNAME
.br
.in +.5i
Get or set the hostname.
.sp
.in -.5i
KERN_JOB_CONTROL
.br
.in +.5i
Return 1 if job control is available on this system, otherwise 0.
.sp
.in -.5i
.\"KERN_LINK_MAX
.\"The maximum file link count.
KERN_MAXFILES
.br
.in +.5i
The maximum number of open files that may be open in the system.
.sp
.in -.5i
KERN_MAXPROC
.br
.in +.5i
The maximum number of simultaneous processes the system will allow.
.sp
.in -.5i
KERN_MAXINODES
.br
.in +.5i
The maximum number of inodes available on the system.
.sp
.in -.5i
KERN_MAXTEXTS
.br
.in +.5i
The maximum number of text structures available on the system.
.sp
.in -.5i
.\"KERN_MAX_CANON
.\"The maximum number of bytes in terminal canonical input line.
.\"KERN_MAX_INPUT
.\"The minimum maximum number of bytes for which space is available in
.\"a terminal input queue.
.\"KERN_NAME_MAX
.\"The maximum number of bytes in a file name.
KERN_NGROUPS
.br
.in +.5i
The maximum number of supplemental groups.
.sp
.in -.5i
.\"KERN_NO_TRUNC
.\"Return 1 if file names longer than KERN_NAME_MAX are truncated.
KERN_OSRELEASE
.br
.in +.5i
The system release string.
.sp
.in -.5i
KERN_OSREV
.br
.in +.5i
The system revision number.
.sp
.in -.5i
KERN_OSTYPE
.br
.in +.5i
The system type string.
.sp
.in -.5i
KERN_OSVERSION
.br
.in +.5i
The kernel build version.
.sp
.in -.5i
.\"KERN_PATH_MAX
.\"The maximum number of bytes in a pathname.
.\"KERN_PIPE_BUF
.\"The maximum number of bytes which will be written atomically to a pipe.
KERN_POSIX1
.br
.in +.5i
The version of ISO/IEC 9945 (POSIX 1003.1) with which the system
attempts to comply.
.sp
.in -.5i
KERN_PROC
.br
.in +.5i
Return the entire process table, or a subset of it.
An array of
\fIstruct kinfo_proc\fP
structures is returned,
whose size depends on the current number of such objects in the system.
.sp
.in -.5i
The third and fourth level names are as follows:
.sp
.nf
.ta .25i 3.0i
	Third level name	Fourth level is:
	KERN\_PROC\_ALL	None
	KERN\_PROC\_PID	A process ID
	KERN\_PROC\_PGRP	A process group
	KERN\_PROC\_TTY	A tty device
	KERN\_PROC\_UID	A user ID

ChangeLog
===========================================================================
2020-10-20
Tuesday, October 20, 2020

Started a ChangeLog file to keep track of my daily progress. I will need to
add in issues I have previously come up with and that which are not yet
included here.

I pushed a commit today of GPIO toggling (for the onboard LEDs) which goes
into great detail of each required part.
There are four registers to configure, in addition to configuring the RCC
clock. There are two ways to set or reset a GPIO pin once it is all set up:
by a read/modify/write dance, or with an atomic bit set or reset operation.

For setting up the GPIO (for port D, for the LEDs), this needs to be done:
 Set up the gpio:
1. clock through the RCC_AHB1ENR register.
2. mode through the GPIOD_MODER register.
3. output as push-pull or open-drain through the GPIOD_OTYPER register.
4. port speed through the GPIOD_OSPEEDR register.
5. pull-up or pull-down resistors through the GPIOD_PUPDR register.

For actually using the GPIO as an output there are two ways:
1. Read/modify/write of the GPIOD_ODR Output Data Register
2. Writing to the GPIOD_BSRR Bit Set/Reset Register

Whichever one is used (or available) is chip vendor dependent. These are
both available for STM32F4 chips.

I put the linker script in the testing base directory (with Makefile.inc)
so that all testing programs can use it, instead of having each program
duplicating the linker script.

The program I finished today has links to all the documentation I have used
to make that program. I will end up putting them in this repository.

I added a .gitignore to the testing directory. Not sure if it is best to
use a global one yet, so just this local one for now.

===========================================================================
2020-10-31
Saturday, October 31, 2020

Pushed two commits that import the RetroBSD repo into this project repo.
I also included the zip from the autobuilder so that there is some starting
point that is easily referenced (instead of just a commit id).


The last while I've been trying to get a mips-elf port of gcc, binutils,
and gdb to build the RetroBSD repo (which is mips-based). I had already got
it to work on Ubuntu (through the Windows Subsystem for Linux) by following
the instructions at http://retrobsd.org/wiki/doku.php/doc/toolchain-mips
which took quite a bit of time to go through.

Also, these instructions are made for an old version of gcc, compiled with
another old version of gcc. The newer gcc's are way more pedantic and find
so many more bugs, even small ones, so it has been a challenge. For
instance, one bug I came across was that glibc changed the location of some
defines into a different header, so I had to find out which one it was and
make a patch for it. It's not hard, just tedious and unfulfilling work.


Get first crack at a devel/mips-elf OpenBSD port done. It has gcc 4.8.1 and
binutils 2.27. Many patches were made. 8 for binutils, 19 for gcc.
I based this port on the ports devel/xtensa-elf and devel/arm-none-eabi.
gdb and newlib are not done yet (and maybe never will be).
binutils was the first I got working. It took a while to get gcc to work.

This port is for compiling the original code of RetroBSD on OpenBSD. The
original code targets the PIC32 MIPS-based microcontroller, thus mips-elf.

There are two or three segfaults in the process of building gcc. I have no
reason why, and I haven't investigated. Starting the build over again
allows a bit more progress until there is another segfault. I just kept
restarting it until the build is complete. The segfaults could be due to
running out of memory or something like that. Who knows. gcc is complex.

There are some post-patch and post-install fixes in the devel/avr/gcc
Makefile that need to be used. This will get rid of all the ridiculous
patches for the single space character needed by c++11 for variables.

===========================================================================
2020-11-01
Sunday, November 01, 2020

I started to build DiscoBSD with my new toolchain, but there are quite a
few errors. The main one is (after much investigation into the cause) due
to gcc not passing -msoft-float to the assembler. This caused the assembler
to assemble as -mhard-float. The linker wasn't able to link these two
separate types of objects.

I fixed this by getting a patchset from gcc, which I had to modify, as it
was meant for some newer version of gcc (5?).

I made 4 new patches, and added some more changes to an already existing
patch. That fixed the whole floating-point issue. The userland programs can
be built now. The kernel builds, too. There still is a problem with the
linker overlaying sections of the executable. I haven't researched it yet,
though.

The VirtualMIPS simulator isn't building. Probably due to not having the
libelf package on OpenBSD. I may need to port that library, too. Who knows.

Other problems are: elf2aout, manpages, init_main.c, expr, yacc kconfig
reduce/reduce conflicts, install not working.

I updated the README.md file for tools/openbsd to include how to get
binutils and gcc for the MIPS target.


Now I am trying to get VirtualMIPS to work. Most errors are about specific
types used. I am guessing they are defined in the libelf code.

I have had to (temporarily) hard-code some directory paths in the Makefile
for VirtualMIPS to what OpenBSD uses, such as /usr/local/include instead of
/opt/local/include on Linux.


Some cool things about the name DiscoBSD:
First, 'disco' is Latin for 'I learn' or 'to learn' or 'become aquainted
with' or 'I study, practice'. All of these definitions are quite in line
with the whole purpose of this directed study, which is to do a heck of a

sbin/sysctl/sysctl.c
/*
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#include <sys/param.h>
#include <sys/stat.h>
#include <sys/sysctl.h>
#include <machine/cpu.h>

#include <netinet/in.h>
#if 0 /* XXX */
#include <sys/socket.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_icmp.h>
#include <netinet/icmp_var.h>
#include <netinet/ip_var.h>
#include <netinet/udp.h>
#include <netinet/udp_var.h>
#endif
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>

struct ctlname topname[] = CTL_NAMES;
struct ctlname kernname[] = CTL_KERN_NAMES;
struct ctlname vmname[] = CTL_VM_NAMES;
#ifdef CTL_NET_NAMES
struct ctlname netname[] = CTL_NET_NAMES;
#endif
struct ctlname hwname[] = CTL_HW_NAMES;
struct ctlname username[] = CTL_USER_NAMES;
struct ctlname debugname[CTL_DEBUG_MAXID];
struct ctlname machdepname[] = CTL_MACHDEP_NAMES;
char names[BUFSIZ];

struct list {
	struct	ctlname *list;
	int	size;
};
struct list toplist = { topname, CTL_MAXID };
struct list secondlevel[] = {
	{ 0, 0 },			/* CTL_UNSPEC */
	{ kernname, KERN_MAXID },	/* CTL_KERN */
	{ vmname, VM_MAXID },		/* CTL_VM */
	{ 0, 0 },			/* CTL_FS */
#ifdef CTL_NET_NAMES
	{ netname, NET_MAXID },		/* CTL_NET */
#else
	{ 0, 0 },
#endif
	{ 0, CTL_DEBUG_MAXID },		/* CTL_DEBUG */
	{ hwname, HW_MAXID },		/* CTL_HW */
	{ machdepname, CPU_MAXID },	/* CTL_MACHDEP */
	{ username, USER_MAXID },	/* CTL_USER_NAMES */
};

void	listall(char *, struct list *);
void	parse(char *, int);
void	debuginit();
#ifdef PF_INET
int	sysctl_inet(char *, char **, int *, int, int *);
#endif /* PF_INET */
int	findname(char *, char *, char **, struct list *);
void	usage();

int	Aflag, aflag, nflag, wflag;

extern char *optarg;
extern int optind, errno;

/*
 * Variables requiring special processing.
 */
#define	CLOCK		0x0001
#define	BOOTTIME	0x0002
#define	CONSDEV		0x0004

int
main(argc, argv)
	int argc;
	char *argv[];
{
	int ch, lvl1;

	while ((ch = getopt(argc, argv, "Aanw")) != EOF) {
		switch (ch) {

		case 'A':
			Aflag = 1;
			break;

		case 'a':
			aflag = 1;
			break;

		case 'n':
			nflag = 1;
			break;

		case 'w':
			wflag = 1;
			break;

		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if (Aflag || aflag) {
		debuginit();
		for (lvl1 = 1; lvl1 < CTL_MAXID; lvl1++)
			listall(topname[lvl1].ctl_name, &secondlevel[lvl1]);
		exit(0);
	}
	if (argc == 0)
		usage();
	while (argc-- > 0)
		parse(*argv, 1);
	exit(0);
}

/*
 * List all variables known to the system.
 */
void
listall(prefix, lp)
	char *prefix;
	struct list *lp;
{
	int lvl2;
	char *cp, name[BUFSIZ];

	if (lp->list == 0)
		return;
	strcpy(name, prefix);
	cp = &name[strlen(name)];
	*cp++ = '.';
	for (lvl2 = 0; lvl2 < lp->size; lvl2++) {
		if (lp->list[lvl2].ctl_name == 0)
			continue;
		strcpy(cp, lp->list[lvl2].ctl_name);
		parse(name, Aflag);
	}
}

/*
 * Parse a name into a MIB entry.
 * Lookup and print out the MIB entry if it exists.
 * Set a new value if requested.
 */
void
parse(string, flags)
	char *string;
	int flags;
{
	int indx, type, state, size, len;
	int special = 0;
	void *newval = (void *)0;
	int intval, newsize = 0;
	long longval;
	struct list *lp;
	int mib[CTL_MAXNAME];
	char *cp, *bufp, buf[BUFSIZ], strval[BUFSIZ];

	bufp = buf;
	strcpy(buf, string);
	if ((cp = strchr(string, '=')) != NULL) {
		if (!wflag) {
			fprintf(stderr, "Must specify -w to set variables\n");
			exit(2);
		}
		*strchr(buf, '=') = '\0';
		*cp++ = '\0';
		while (isspace(*cp))
			cp++;
		newval = (void *)cp;
		newsize = strlen(cp);
	}
	if ((indx = findname(string, "top", &bufp, &toplist)) == -1)
		return;
	mib[0] = indx;
	if (indx == CTL_DEBUG)
		debuginit();
	lp = &secondlevel[indx];
	if (lp->list == 0) {
		fprintf(stderr, "%s: class is not implemented\n",
		    topname[indx]);
		return;
	}
	if (bufp == NULL) {
		listall(topname[indx].ctl_name, lp);
		return;
	}
	if ((indx = findname(string, "second", &bufp, lp)) == -1)
		return;
	mib[1] = indx;
	type = lp->list[indx].ctl_type;
	len = 2;
	switch (mib[0]) {

	case CTL_KERN:
		switch (mib[1]) {
		case KERN_PROF:
			if (flags == 0)
				return;
			fprintf(stderr,
			    "%s is not supported in DiscoBSD\n", string);
			return;
		case KERN_INODE:
		case KERN_FILE:
		case KERN_TEXT:
			if (flags == 0)
				return;
			fprintf(stderr,
			    "Use pstat to view %s information\n", string);
			return;
		case KERN_PROC:
			if (flags == 0)
				return;
			fprintf(stderr,
			    "Use ps to view %s information\n", string);
			return;
		case KERN_CLOCKRATE:
			special |= CLOCK;
			break;
		case KERN_BOOTTIME:
			special |= BOOTTIME;
			break;
		}
		break;

	case CTL_HW:
		break;

	case CTL_VM:
		if (mib[1] == VM_LOADAVG) {
			unsigned loads[3];

			getloadavg(loads, 3);
			if (!nflag)
				fprintf(stdout, "%s = ", string);
			fprintf(stdout, "%u.%02u %u.%02u %u.02u\n",
                                loads[0] / 100, loads[0] % 100,
                                loads[1] / 100, loads[1] % 100,
                                loads[2] / 100, loads[2] % 100);
			return;
		}
		if (flags == 0)
			return;
		fprintf(stderr,
		    "Use vmstat or pstat to view %s information\n", string);
		return;

	case CTL_NET:
#ifdef PF_INET
		if (mib[1] == PF_INET) {
			len = sysctl_inet(string, &bufp, mib, flags, &type);
			if (len >= 0)
				break;
			return;
		}
#endif /* PF_INET */
		if (flags == 0)
			return;
		fprintf(stderr, "Use netstat to view %s information\n", string);
		return;

	case CTL_DEBUG:
		mib[2] = CTL_DEBUG_VALUE;
		len = 3;
		break;

	case CTL_MACHDEP:
		if (mib[1] == CPU_CONSDEV)
			special |= CONSDEV;
		break;

	case CTL_FS:
	case CTL_USER:
		break;

	default:
		fprintf(stderr, "Illegal top level value: %d\n", mib[0]);
		return;

	}
doit:
	if (bufp) {
		fprintf(stderr, "name %s in %s is unknown\n", *bufp, string);
		return;
	}
	if (newsize > 0) {
		switch (type) {
		case CTLTYPE_INT:
			intval = atoi(newval);
			newval = (void *)&intval;
			newsize = sizeof intval;
			break;

		case CTLTYPE_LONG:
			sscanf(newval, "%ld", &longval);
			newval = (void *)&longval;
			newsize = sizeof longval;
			break;
		}
	}
	size = BUFSIZ;
	if (sysctl(mib, len, buf, &size, newsize ? newval : (void *)0, newsize) == -1) {
		if (flags == 0)
			return;
		switch (errno) {
		case EOPNOTSUPP:
			fprintf(stderr, "%s: value is not available\n", string);
			return;
		case ENOTDIR:
			fprintf(stderr, "%s: specification is incomplete\n",
			    string);
			return;
		case ENOMEM:
			fprintf(stderr, "%s: type is unknown to this program\n",
			    string);
			return;
		default:
			perror(string);
			return;
		}
	}
	if (special & CLOCK) {
		struct clockinfo *clkp = (struct clockinfo *)buf;

		if (!nflag)
			fprintf(stdout, "%s = ", string);
		fprintf(stdout,
		    "hz = %d, tick = %d, profhz = %d, stathz = %d\n",
		    clkp->hz, clkp->tick, clkp->profhz, clkp->stathz);
		return;
	}
	if (special & BOOTTIME) {
		struct timeval *btp = (struct timeval *)buf;

		if (!nflag)
			fprintf(stdout, "%s = %s", string,
			    ctime(&btp->tv_sec));
		else
			fprintf(stdout, "%d\n", btp->tv_sec);
		return;
	}
	if (special & CONSDEV) {
		dev_t dev = *(dev_t *)buf;

		if (!nflag)
			fprintf(stdout, "%s = %s\n", string,
			    devname(dev, S_IFCHR));
		else
			fprintf(stdout, "0x%x\n", dev);
		return;
	}
	switch (type) {
	case CTLTYPE_INT:
		if (newsize == 0) {
			if (!nflag)
				fprintf(stdout, "%s = ", string);
			fprintf(stdout, "%d\n", *(int *)buf);
		} else {
			if (!nflag)
				fprintf(stdout, "%s: %d -> ", string,
				    *(int *)buf);
			fprintf(stdout, "%d\n", *(int *)newval);
		}
		return;

	case CTLTYPE_STRING:
		if (newsize == 0) {
			if (!nflag)
				fprintf(stdout, "%s = ", string);
			fprintf(stdout, "%s\n", buf);
		} else {
			if (!nflag)
				fprintf(stdout, "%s: %s -> ", string, buf);
			fprintf(stdout, "%s\n", newval);
		}
		return;

	case CTLTYPE_LONG:
		if (newsize == 0) {
			if (!nflag)
				fprintf(stdout, "%s = ", string);
			fprintf(stdout, "%ld\n", *(long *)buf);
		} else {
			if (!nflag)
				fprintf(stdout, "%s: %ld -> ", string,
				    *(long *)buf);
			fprintf(stdout, "%ld\n", *(long *)newval);
		}
		return;

	case CTLTYPE_STRUCT:
		fprintf(stderr, "%s: unknown structure returned\n",
		    string);
		return;

	default:
	case CTLTYPE_NODE:
		fprintf(stderr, "%s: unknown type returned\n",
		    string);
		return;
	}
}

/*
 * Initialize the set of debugging names
 */
void
debuginit()
{
	int mib[3], size, loc, i;

	if (secondlevel[CTL_DEBUG].list != 0)
		return;
	secondlevel[CTL_DEBUG].list = debugname;
	mib[0] = CTL_DEBUG;
	mib[2] = CTL_DEBUG_NAME;
	for (loc = 0, i = 0; i < CTL_DEBUG_MAXID; i++) {
		mib[1] = i;
		size = BUFSIZ - loc;
		if (sysctl(mib, 3, &names[loc], &size, NULL, 0) == -1)
			continue;
		debugname[i].ctl_name = &names[loc];
		debugname[i].ctl_type = CTLTYPE_INT;
		loc += size;
	}
}

#ifdef PF_INET
struct ctlname inetname[] = CTL_IPPROTO_NAMES;
struct ctlname ipname[] = IPCTL_NAMES;
struct ctlname icmpname[] = ICMPCTL_NAMES;
struct ctlname udpname[] = UDPCTL_NAMES;
struct list inetlist = { inetname, IPPROTO_MAXID };
struct list inetvars[] = {
	{ ipname, IPCTL_MAXID },	/* ip */
	{ icmpname, ICMPCTL_MAXID },	/* icmp */
	{ 0, 0 },			/* igmp */
	{ 0, 0 },			/* ggmp */
	{ 0, 0 },
	{ 0, 0 },
	{ 0, 0 },			/* tcp */
	{ 0, 0 },
	{ 0, 0 },			/* egp */
	{ 0, 0 },
	{ 0, 0 },
	{ 0, 0 },
	{ 0, 0 },			/* pup */
	{ 0, 0 },
	{ 0, 0 },
	{ 0, 0 },
	{ 0, 0 },
	{ udpname, UDPCTL_MAXID },	/* udp */
};

/*
 * handle internet requests
 */
int
sysctl_inet(string, bufpp, mib, flags, typep)
	char *string;
	char **bufpp;
	int mib[];
	int flags;
	int *typep;
{
	struct list *lp;
	int indx;

	if (*bufpp == NULL) {
		listall(string, &inetlist);
		return (-1);
	}
	if ((indx = findname(string, "third", bufpp, &inetlist)) == -1)
		return (-1);
	mib[2] = indx;
	if (indx <= IPPROTO_UDP && inetvars[indx].list != NULL)
		lp = &inetvars[indx];
	else if (!flags)
		return (-1);
	else {
		fprintf(stderr, "%s: no variables defined for this protocol\n",
		    string);
		return (-1);
	}
	if (*bufpp == NULL) {
		listall(string, lp);
		return (-1);
	}
	if ((indx = findname(string, "fourth", bufpp, lp)) == -1)
		return (-1);
	mib[3] = indx;
	*typep = lp->list[indx].ctl_type;
	return (4);
}
#endif /* PF_INET */

/*
 * Scan a list of names searching for a particular name.
 */
int
findname(string, level, bufp, namelist)
	char *string;
	char *level;
	char **bufp;
	struct list *namelist;
{
	char *name;
	int i;

	if (namelist->list == 0 || (name = strsep(bufp, ".")) == NULL) {
		fprintf(stderr, "%s: incomplete specification\n", string);
		return (-1);
	}
	for (i = 0; i < namelist->size; i++)
		if (namelist->list[i].ctl_name != NULL &&
		    strcmp(name, namelist->list[i].ctl_name) == 0)
			break;
	if (i == namelist->size) {
		fprintf(stderr, "%s level name %s in %s is invalid\n",
		    level, name, string);
		return (-1);
	}
	return (i);
}

void
usage()
{

	(void)fprintf(stderr, "usage:\t%s\n\t%s\n\t%s\n\t%s\n",
	    "sysctl [-n] variable ...", "sysctl [-n] -w variable=value ...",
	    "sysctl [-n] -a", "sysctl [-n] -A");
	exit(1);
}
